### ISSUE TITLE
Standardize self_link handling & customization

### ISSUE THREAD

**Issue by melinath** [hashicorp/terraform-provider-google#20109]
### What kind of contribution is this issue about?

MMv1-based resource

### Details

There are a number of API resources that use GCE self_links (which can have full https version or a number of different "short" variants), but we don't have great tools to make them behave in a standard way; people have to reinvent the behavior or infer it from other resources. We should take the following steps:

- Make a SelfLink field type in MMv1 with customizable behavior
- Eliminate all custom templates that work with self_links:
  - self_link_from_name
  - resource_from_self_link
  - name_from_self_link
  - etc.

### References

https://github.com/GoogleCloudPlatform/magic-modules/pull/12111#discussion_r1823121983

**Comment by melinath** [comment]
Yeah, discovering it's just not possible could be the outcome here.

**Comment by c2thorn** [comment]
Would probably need to break the investigation off from this.

**Comment by c2thorn** [comment]
May be shared and safe within GCE, but not safe across all other products. 

This is a large task and the investigation needs to scope this down and suggest and justify potential solutions

We should try to break this down into smaller fixit-size items to reduce the scope. We are likely to never do this as-is
